File: src/vendors/wc3maptranslator/translators/MapTranslator.ts
import { existsSync, readFileSync, writeFileSync } from 'fs';
import path from 'path';

import {
  Camera, Doodad, Info, ObjectModificationTable, ObjectType, Region, SpecialDoodad, Terrain, Unit,
} from '../data';
import { CamerasTranslator } from './CamerasTranslator';
import { DoodadsTranslator } from './DoodadsTranslator';
import { InfoTranslator } from './InfoTranslator';
import { ObjectsTranslator } from './ObjectsTranslator';
import { RegionsTranslator } from './RegionsTranslator';
import { TerrainTranslator } from './TerrainTranslator';
import { UnitsTranslator } from './UnitsTranslator';

type FilePath = 'info'
  | 'terrain'
  | 'units'
  | 'doodads'
  | 'cameras'
  | 'regions'
  | 'unitData'
  | 'unitTypeSkins'
  | 'destructibleTypeSkins'
  | 'doodadTypeSkins'
  | 'abilityTypeSkins'
  | 'buffTypeSkins'
  | 'itemData'
  | 'destructibleData'
  | 'doodadData'
  | 'abilityData'
  | 'buffData'
  | 'upgradeData';

export class MapTranslator {
  public info: Info;

  public terrain: Terrain;

  public units: Unit[] = [];

  public doodads: Doodad[] = [];

  public specialDoodads: SpecialDoodad[] = [];

  public cameras: Camera[] = [];

  public regions: Region[] = [];

  public unitData: ObjectModificationTable = { original: {}, custom: {} };

  public destructibleData: ObjectModificationTable = { original: {}, custom: {} };

  public doodadData: ObjectModificationTable = { original: {}, custom: {} };

  public itemData: ObjectModificationTable = { original: {}, custom: {} };

  public abilityData: ObjectModificationTable = { original: {}, custom: {} };

  public buffData: ObjectModificationTable = { original: {}, custom: {} };

  public upgradeData: ObjectModificationTable = { original: {}, custom: {} };

  public unitTypeSkins: ObjectModificationTable = { original: {}, custom: {} };

  public destructibleTypeSkins: ObjectModificationTable = { original: {}, custom: {} };

  public doodadTypeSkins: ObjectModificationTable = { original: {}, custom: {} };

  public abilityTypeSkins: ObjectModificationTable = { original: {}, custom: {} };

  public buffTypeSkins: ObjectModificationTable = { original: {}, custom: {} };

  public filePaths: Record<FilePath, string>;

  constructor() {
  }

  public setMapDir(mapDir: string) {
    this.filePaths = {
      info: path.join(mapDir, 'war3map.w3i'),
      terrain: path.join(mapDir, 'war3map.w3e'),
      units: path.join(mapDir, 'war3mapUnits.doo'),
      doodads: path.join(mapDir, 'war3map.doo'),
      cameras: path.join(mapDir, 'war3map.w3c'),
      regions: path.join(mapDir, 'war3map.w3r'),
      unitData: path.join(mapDir, 'war3map.w3u'),
      unitTypeSkins: path.join(mapDir, 'war3mapSkin.w3u'),
      destructibleTypeSkins: path.join(mapDir, 'war3mapSkin.w3b'),
      doodadTypeSkins: path.join(mapDir, 'war3mapSkin.w3d'),
      abilityTypeSkins: path.join(mapDir, 'war3mapSkin.w3a'),
      buffTypeSkins: path.join(mapDir, 'war3mapSkin.w3h'),
      itemData: path.join(mapDir, 'war3map.w3t'),
      destructibleData: path.join(mapDir, 'war3map.w3b'),
      doodadData: path.join(mapDir, 'war3map.w3d'),
      abilityData: path.join(mapDir, 'war3map.w3a'),
      buffData: path.join(mapDir, 'war3map.w3h'),
      upgradeData: path.join(mapDir, 'war3map.w3q'),
    };
  }

  load(mapDir: string) {
    this.setMapDir(mapDir);
    this.info = InfoTranslator.warToJson(readFileSync(this.filePaths.info)).json;
    this.terrain = TerrainTranslator.warToJson(readFileSync(this.filePaths.terrain)).json;

    console.log('Loading units from', this.filePaths.unitData);
    this.units = UnitsTranslator.warToJson(readFileSync(this.filePaths.units)).json;
    console.log('Units loaded', this.units.length);

    const allDoodads = DoodadsTranslator.warToJson(readFileSync(this.filePaths.doodads)).json;
    this.doodads = allDoodads[0];
    this.specialDoodads = allDoodads[1];

    if (existsSync(this.filePaths.cameras)) {
      this.cameras = CamerasTranslator.warToJson(readFileSync(this.filePaths.cameras)).json;
    }

    if (existsSync(this.filePaths.regions)) {
      this.regions = RegionsTranslator.warToJson(readFileSync(this.filePaths.regions)).json;
    }

    if (existsSync(this.filePaths.unitData)) {
      this.unitData = ObjectsTranslator.warToJson(ObjectType.Units, readFileSync(this.filePaths.unitData)).json;
    }
    if (existsSync(this.filePaths.unitTypeSkins)) {
      this.unitTypeSkins = ObjectsTranslator.warToJson(ObjectType.Units, readFileSync(this.filePaths.unitTypeSkins)).json;
    }
    if (existsSync(this.filePaths.destructibleTypeSkins)) {
      this.destructibleTypeSkins = ObjectsTranslator.warToJson(ObjectType.Destructables, readFileSync(this.filePaths.destructibleTypeSkins)).json;
    }
    if (existsSync(this.filePaths.doodadTypeSkins)) {
      this.doodadTypeSkins = ObjectsTranslator.warToJson(ObjectType.Doodads, readFileSync(this.filePaths.doodadTypeSkins)).json;
    }
    if (existsSync(this.filePaths.abilityTypeSkins)) {
      this.abilityTypeSkins = ObjectsTranslator.warToJson(ObjectType.Abilities, readFileSync(this.filePaths.abilityTypeSkins)).json;
    }
    if (existsSync(this.filePaths.buffTypeSkins)) {
      this.buffTypeSkins = ObjectsTranslator.warToJson(ObjectType.Buffs, readFileSync(this.filePaths.buffTypeSkins)).json;
    }
    if (existsSync(this.filePaths.itemData)) {
      this.itemData = ObjectsTranslator.warToJson(ObjectType.Items, readFileSync(this.filePaths.itemData)).json;
    }
    if (existsSync(this.filePaths.destructibleData)) {
      this.destructibleData = ObjectsTranslator.warToJson(ObjectType.Destructables, readFileSync(this.filePaths.destructibleData)).json;
    }
    if (existsSync(this.filePaths.doodadData)) {
      this.doodadData = ObjectsTranslator.warToJson(ObjectType.Doodads, readFileSync(this.filePaths.doodadData)).json;
    }
    if (existsSync(this.filePaths.abilityData)) {
      this.abilityData = ObjectsTranslator.warToJson(ObjectType.Abilities, readFileSync(this.filePaths.abilityData)).json;
    }
    if (existsSync(this.filePaths.buffData)) {
      this.buffData = ObjectsTranslator.warToJson(ObjectType.Buffs, readFileSync(this.filePaths.buffData)).json;
    }
    if (existsSync(this.filePaths.upgradeData)) {
      this.upgradeData = ObjectsTranslator.warToJson(ObjectType.Upgrades, readFileSync(this.filePaths.upgradeData)).json;
    }
  }

  public save(filePath: FilePath) {
    switch (filePath) {
      case 'info':
        writeFileSync(this.filePaths.info, InfoTranslator.jsonToWar(this.info).buffer);
        break;
      case 'terrain':
        writeFileSync(this.filePaths.terrain, TerrainTranslator.jsonToWar(this.terrain).buffer);
        break;
      case 'units':
        writeFileSync(this.filePaths.units, UnitsTranslator.jsonToWar(this.units).buffer);
        break;
      case 'doodads':
        writeFileSync(this.filePaths.doodads, DoodadsTranslator.jsonToWar([this.doodads, this.specialDoodads]).buffer);
        break;
      case 'cameras':
        writeFileSync(this.filePaths.cameras, CamerasTranslator.jsonToWar(this.cameras).buffer);
        break;
      case 'regions':
        writeFileSync(this.filePaths.regions, RegionsTranslator.jsonToWar(this.regions).buffer);
        break;
      case 'unitData':
        writeFileSync(this.filePaths.unitData, ObjectsTranslator.jsonToWar(ObjectType.Units, this.unitData).buffer);
        break;
      case 'unitTypeSkins':
        writeFileSync(this.filePaths.unitTypeSkins, ObjectsTranslator.jsonToWar(ObjectType.Units, this.unitTypeSkins).buffer);
        break;
      case 'destructibleTypeSkins':
        writeFileSync(this.filePaths.destructibleTypeSkins, ObjectsTranslator.jsonToWar(ObjectType.Destructables, this.destructibleTypeSkins).buffer);
        break;
      case 'doodadTypeSkins':
        writeFileSync(this.filePaths.doodadTypeSkins, ObjectsTranslator.jsonToWar(ObjectType.Doodads, this.doodadTypeSkins).buffer);
        break;
      case 'abilityTypeSkins':
        writeFileSync(this.filePaths.abilityTypeSkins, ObjectsTranslator.jsonToWar(ObjectType.Abilities, this.abilityTypeSkins).buffer);
        break;
      case 'buffTypeSkins':
        writeFileSync(this.filePaths.buffTypeSkins, ObjectsTranslator.jsonToWar(ObjectType.Buffs, this.buffTypeSkins).buffer);
        break;
      case 'itemData':
        writeFileSync(this.filePaths.itemData, ObjectsTranslator.jsonToWar(ObjectType.Items, this.itemData).buffer);
        break;
      case 'destructibleData':
        writeFileSync(this.filePaths.destructibleData, ObjectsTranslator.jsonToWar(ObjectType.Destructables, this.destructibleData).buffer);
        break;
      case 'doodadData':
        writeFileSync(this.filePaths.doodadData, ObjectsTranslator.jsonToWar(ObjectType.Doodads, this.doodadData).buffer);
        break;
      case 'abilityData':
        writeFileSync(this.filePaths.abilityData, ObjectsTranslator.jsonToWar(ObjectType.Abilities, this.abilityData).buffer);
        break;
      case 'buffData':
        writeFileSync(this.filePaths.buffData, ObjectsTranslator.jsonToWar(ObjectType.Buffs, this.buffData).buffer);
        break;
      case 'upgradeData':
        writeFileSync(this.filePaths.upgradeData, ObjectsTranslator.jsonToWar(ObjectType.Upgrades, this.upgradeData).buffer);
        break;
      default:
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        throw new Error(`Unknown file path ${filePath}`);
    }
  }
}

File: golang/internal/wc3/translators/map_translator.go
package translators

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/pqhuy98/wow-converter/internal/wc3/data"
)

// FilePath identifies a war3map file slot.
type FilePath string

const (
	FileInfo                  FilePath = "info"
	FileTerrain               FilePath = "terrain"
	FileUnits                 FilePath = "units"
	FileDoodads               FilePath = "doodads"
	FileCameras               FilePath = "cameras"
	FileRegions               FilePath = "regions"
	FileUnitData              FilePath = "unitData"
	FileUnitTypeSkins         FilePath = "unitTypeSkins"
	FileDestructibleTypeSkins FilePath = "destructibleTypeSkins"
	FileDoodadTypeSkins       FilePath = "doodadTypeSkins"
	FileAbilityTypeSkins      FilePath = "abilityTypeSkins"
	FileBuffTypeSkins         FilePath = "buffTypeSkins"
	FileItemData              FilePath = "itemData"
	FileDestructibleData      FilePath = "destructibleData"
	FileDoodadData             FilePath = "doodadData"
	FileAbilityData           FilePath = "abilityData"
	FileBuffData              FilePath = "buffData"
	FileUpgradeData           FilePath = "upgradeData"
)

// MapTranslator loads and saves WC3 map binary files.
type MapTranslator struct {
	Info                  data.Info
	Terrain               data.Terrain
	Units                 []data.Unit
	Doodads               []data.Doodad
	SpecialDoodads        []data.SpecialDoodad
	Cameras               []data.Camera
	Regions               []data.Region
	UnitData              data.ObjectModificationTable
	DestructibleData      data.ObjectModificationTable
	DoodadData            data.ObjectModificationTable
	ItemData              data.ObjectModificationTable
	AbilityData           data.ObjectModificationTable
	BuffData              data.ObjectModificationTable
	UpgradeData           data.ObjectModificationTable
	UnitTypeSkins         data.ObjectModificationTable
	DestructibleTypeSkins data.ObjectModificationTable
	DoodadTypeSkins       data.ObjectModificationTable
	AbilityTypeSkins      data.ObjectModificationTable
	BuffTypeSkins         data.ObjectModificationTable
	filePaths             map[FilePath]string
}

// NewMapTranslator returns an empty map translator.
func NewMapTranslator() *MapTranslator {
	return &MapTranslator{
		UnitData:              data.NewObjectModificationTable(),
		DestructibleData:      data.NewObjectModificationTable(),
		DoodadData:            data.NewObjectModificationTable(),
		ItemData:              data.NewObjectModificationTable(),
		AbilityData:           data.NewObjectModificationTable(),
		BuffData:              data.NewObjectModificationTable(),
		UpgradeData:           data.NewObjectModificationTable(),
		UnitTypeSkins:         data.NewObjectModificationTable(),
		DestructibleTypeSkins: data.NewObjectModificationTable(),
		DoodadTypeSkins:       data.NewObjectModificationTable(),
		AbilityTypeSkins:      data.NewObjectModificationTable(),
		BuffTypeSkins:         data.NewObjectModificationTable(),
	}
}

// SetMapDir configures output paths for map files.
func (m *MapTranslator) SetMapDir(mapDir string) {
	m.filePaths = map[FilePath]string{
		FileInfo:                  filepath.Join(mapDir, "war3map.w3i"),
		FileTerrain:               filepath.Join(mapDir, "war3map.w3e"),
		FileUnits:                 filepath.Join(mapDir, "war3mapUnits.doo"),
		FileDoodads:               filepath.Join(mapDir, "war3map.doo"),
		FileCameras:               filepath.Join(mapDir, "war3map.w3c"),
		FileRegions:               filepath.Join(mapDir, "war3map.w3r"),
		FileUnitData:              filepath.Join(mapDir, "war3map.w3u"),
		FileDestructibleData:      filepath.Join(mapDir, "war3map.w3b"),
		FileDoodadData:            filepath.Join(mapDir, "war3map.w3d"),
		FileItemData:              filepath.Join(mapDir, "war3map.w3t"),
		FileAbilityData:           filepath.Join(mapDir, "war3map.w3a"),
		FileBuffData:              filepath.Join(mapDir, "war3map.w3h"),
		FileUpgradeData:           filepath.Join(mapDir, "war3map.w3q"),
		FileUnitTypeSkins:         filepath.Join(mapDir, "war3mapSkin.w3u"),
		FileDestructibleTypeSkins: filepath.Join(mapDir, "war3mapSkin.w3b"),
		FileDoodadTypeSkins:       filepath.Join(mapDir, "war3mapSkin.w3d"),
		FileAbilityTypeSkins:     filepath.Join(mapDir, "war3mapSkin.w3a"),
		FileBuffTypeSkins:         filepath.Join(mapDir, "war3mapSkin.w3h"),
	}
}

// Load reads map files from mapDir.
func (m *MapTranslator) Load(mapDir string) error {
	m.SetMapDir(mapDir)

	infoBytes, err := os.ReadFile(m.filePaths[FileInfo])
	if err != nil {
		return fmt.Errorf("load map %s: %w", mapDir, err)
	}
	m.Info = InfoTranslator{}.WarToJSON(infoBytes).JSON

	terrainBytes, err := os.ReadFile(m.filePaths[FileTerrain])
	if err != nil {
		return fmt.Errorf("load terrain: %w", err)
	}
	m.Terrain = TerrainTranslator{}.WarToJSON(terrainBytes).JSON

	unitsBytes, err := os.ReadFile(m.filePaths[FileUnits])
	if err != nil {
		return fmt.Errorf("load units: %w", err)
	}
	m.Units = UnitsTranslator{}.WarToJSON(unitsBytes).JSON

	doodadsBytes, err := os.ReadFile(m.filePaths[FileDoodads])
	if err != nil {
		return fmt.Errorf("load doodads: %w", err)
	}
	allDoodads := DoodadsTranslator{}.WarToJSON(doodadsBytes).JSON
	m.Doodads = allDoodads.Doodads
	m.SpecialDoodads = allDoodads.SpecialDoodads

	if _, err := os.Stat(m.filePaths[FileCameras]); err == nil {
		camerasBytes, readErr := os.ReadFile(m.filePaths[FileCameras])
		if readErr != nil {
			return fmt.Errorf("load cameras: %w", readErr)
		}
		m.Cameras = CamerasTranslator{}.WarToJSON(camerasBytes).JSON
	}

	if _, err := os.Stat(m.filePaths[FileRegions]); err == nil {
		regionsBytes, readErr := os.ReadFile(m.filePaths[FileRegions])
		if readErr != nil {
			return fmt.Errorf("load regions: %w", readErr)
		}
		m.Regions = RegionsTranslator{}.WarToJSON(regionsBytes).JSON
	}

	m.loadObjectFile(FileUnitData, data.ObjectUnits, &m.UnitData)
	m.loadObjectFile(FileUnitTypeSkins, data.ObjectUnits, &m.UnitTypeSkins)
	m.loadObjectFile(FileDestructibleTypeSkins, data.ObjectDestructables, &m.DestructibleTypeSkins)
	m.loadObjectFile(FileDoodadTypeSkins, data.ObjectDoodads, &m.DoodadTypeSkins)
	m.loadObjectFile(FileAbilityTypeSkins, data.ObjectAbilities, &m.AbilityTypeSkins)
	m.loadObjectFile(FileBuffTypeSkins, data.ObjectBuffs, &m.BuffTypeSkins)
	m.loadObjectFile(FileItemData, data.ObjectItems, &m.ItemData)
	m.loadObjectFile(FileDestructibleData, data.ObjectDestructables, &m.DestructibleData)
	m.loadObjectFile(FileDoodadData, data.ObjectDoodads, &m.DoodadData)
	m.loadObjectFile(FileAbilityData, data.ObjectAbilities, &m.AbilityData)
	m.loadObjectFile(FileBuffData, data.ObjectBuffs, &m.BuffData)
	m.loadObjectFile(FileUpgradeData, data.ObjectUpgrades, &m.UpgradeData)

	return nil
}

func (m *MapTranslator) loadObjectFile(file FilePath, objectType data.ObjectType, dest *data.ObjectModificationTable) {
	if _, err := os.Stat(m.filePaths[file]); err != nil {
		return
	}
	bytes, err := os.ReadFile(m.filePaths[file])
	if err != nil {
		return
	}
	*dest = WarToJSONObjects(objectType, bytes).JSON
}

// Save writes a single map file.
func (m *MapTranslator) Save(file FilePath) error {
	path, ok := m.filePaths[file]
	if !ok || path == "" {
		return fmt.Errorf("map dir not set")
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}

	var out []byte
	switch file {
	case FileInfo:
		out = InfoTranslator{}.JSONToWar(m.Info).Buffer
	case FileTerrain:
		out = TerrainTranslator{}.JSONToWar(m.Terrain).Buffer
	case FileUnits:
		out = UnitsTranslator{}.JSONToWar(m.Units).Buffer
	case FileDoodads:
		out = DoodadsTranslator{}.JSONToWar(data.DoodadList{
			Doodads: m.Doodads, SpecialDoodads: m.SpecialDoodads,
		}).Buffer
	case FileCameras:
		out = CamerasTranslator{}.JSONToWar(m.Cameras).Buffer
	case FileRegions:
		out = RegionsTranslator{}.JSONToWar(m.Regions).Buffer
	case FileUnitData:
		out = JSONToWarObjects(data.ObjectUnits, m.UnitData).Buffer
	case FileUnitTypeSkins:
		out = JSONToWarObjects(data.ObjectUnits, m.UnitTypeSkins).Buffer
	case FileDestructibleTypeSkins:
		out = JSONToWarObjects(data.ObjectDestructables, m.DestructibleTypeSkins).Buffer
	case FileDoodadTypeSkins:
		out = JSONToWarObjects(data.ObjectDoodads, m.DoodadTypeSkins).Buffer
	case FileAbilityTypeSkins:
		out = JSONToWarObjects(data.ObjectAbilities, m.AbilityTypeSkins).Buffer
	case FileBuffTypeSkins:
		out = JSONToWarObjects(data.ObjectBuffs, m.BuffTypeSkins).Buffer
	case FileItemData:
		out = JSONToWarObjects(data.ObjectItems, m.ItemData).Buffer
	case FileDestructibleData:
		out = JSONToWarObjects(data.ObjectDestructables, m.DestructibleData).Buffer
	case FileDoodadData:
		out = JSONToWarObjects(data.ObjectDoodads, m.DoodadData).Buffer
	case FileAbilityData:
		out = JSONToWarObjects(data.ObjectAbilities, m.AbilityData).Buffer
	case FileBuffData:
		out = JSONToWarObjects(data.ObjectBuffs, m.BuffData).Buffer
	case FileUpgradeData:
		out = JSONToWarObjects(data.ObjectUpgrades, m.UpgradeData).Buffer
	default:
		return fmt.Errorf("unknown file path %s", file)
	}
	return os.WriteFile(path, out, 0o644)
}
